pkg/controller/dual-pods/server-requesting.go

```diff
@@ -0,0 +1,229 @@
+		return err, true
+	}
+	logger.V(5).Info("Created server-running pod", "name", serverRunningPod.Name)
+
+	logger.V(5).Info("Processed server-requesting pod", "name", requestingPod.Name)
+	return nil, false
+}
+
+// serverTemplateData defines values available to the server-running pod's template.
+type serverTemplateData struct {
+	GPUIndices   string
+	LocalVolume  string
+	SharedVolume string
+}
+
+func composeServerRunningPod(reqPod *corev1.Pod, data serverTemplateData) (*corev1.Pod, error) {
+	rawTmpl, ok := reqPod.Annotations[api.ServerPatchAnnotationName]
+	if !ok {
+		return nil, fmt.Errorf("annotation %q not found", api.ServerPatchAnnotationName)
+	}
+
+	tmpl, err := template.New("serverPatch").Option("missingkey=error").Parse(rawTmpl)
+	if err != nil {
+		return nil, fmt.Errorf("parse template: %w", err)
+	}
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return nil, fmt.Errorf("execute template: %w", err)
+	}
+	renderedPatch := buf.Bytes()
+
+	patchJSON, err := yaml.YAMLToJSON(renderedPatch)
+	if err != nil {
+		return nil, fmt.Errorf("yaml to json: %w", err)
+	}
+
+	reqPod.Spec.Containers = []corev1.Container{}
+	reqPod.Spec.InitContainers = nil
+	origJSON, err := json.Marshal(reqPod)
+	if err != nil {
+		return nil, fmt.Errorf("marshal server-requesting pod: %w", err)
+	}
+
+	// apply strategic merge patch
+	modifiedJSON, err := strategicpatch.StrategicMergePatch(origJSON, patchJSON, &corev1.Pod{})
+	if err != nil {
+		return nil, fmt.Errorf("apply patch: %w", err)
+	}
+
+	// decode back into Pod
+	var pod corev1.Pod
+	if err := json.Unmarshal(modifiedJSON, &pod); err != nil {
+		return nil, fmt.Errorf("unmarshal patched pod: %w", err)
+	}
+
+	// ensure the correct role annotation
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+	pod.Annotations[api.PodRoleAnnotationName] = api.PodRoleAnnotationValueRunning
+
+	// set the inference server container's gpu limits and requests to zero to bypass the nvidia device plugin
+	for i, c := range pod.Spec.Containers {
+		if c.Name != api.InferenceServerContainerName {
+			continue
+		}
+		if pod.Spec.Containers[i].Resources.Limits == nil {
+			pod.Spec.Containers[i].Resources.Limits = corev1.ResourceList{}
+		}
+		pod.Spec.Containers[i].Resources.Limits[corev1.ResourceName("nvidia.com/gpu")] = resource.Quantity{}
+		if pod.Spec.Containers[i].Resources.Requests == nil {
+			pod.Spec.Containers[i].Resources.Requests = corev1.ResourceList{}
+		}
+		pod.Spec.Containers[i].Resources.Requests[corev1.ResourceName("nvidia.com/gpu")] = resource.Quantity{}
+	}
+
+	// connect dual pods
+	pod.Name = reqPod.Name + api.ServerRunningPodNameSuffix
+	pod.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(reqPod, corev1.SchemeGroupVersion.WithKind("Pod")),
+	}
+
+	// clean up
+	delete(pod.Annotations, api.AdminPortAnnotationName)
+	delete(pod.Annotations, api.ServerPatchAnnotationName)
+	pod.ResourceVersion = ""
+	pod.UID = ""
```

Comment:
You should review all of https://github.com/kubernetes/apimachinery/blob/v0.33.5/pkg/apis/meta/v1/types.go#L111-L276